lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/CollateralSpec.hs

```diff
@@ -0,0 +1,810 @@
+            addrType = B.runGet getHeader (BL.fromStrict addrBytes)
+        disjoin
+          [ addrType === StakeAddress CredentialKeyHash
+          , addrType === StakeAddress CredentialScriptHash
+          ]
+
+-- | Test that for any shelley keyhash address, we classify it as a shelley
+-- keyhash address (although not necessarily the correct one, as it's a bit
+-- difficult to assert that the exact type is correct).
+prop_addressType_shelleyKeyHash :: Property
+prop_addressType_shelleyKeyHash =
+    forAll genShelleyKeyHashAddr $ \shelleyKeyHashAddr -> do
+        let
+            (Address addrBytes) = asAddress shelleyKeyHashAddr
+            addrType = B.runGet getHeader (BL.fromStrict addrBytes)
+        disjoin
+          [ addrType === BaseAddress CredentialKeyHash CredentialKeyHash
+          , addrType === BaseAddress CredentialKeyHash CredentialScriptHash
+          , addrType === PointerAddress CredentialKeyHash
+          , addrType === EnterpriseAddress CredentialKeyHash
+          ]
+
+-- | Test that for any shelley scripthash address, we classify it as a shelley
+-- scripthash address (although not necessarily the correct one, as it's a bit
+-- difficult to assert that the exact type is correct).
+prop_addressType_shelleyScriptHash :: Property
+prop_addressType_shelleyScriptHash =
+    forAll genShelleyScriptHashAddr $ \shelleyScriptHashAddr -> do
+        let
+            (Address addrBytes) = asAddress shelleyScriptHashAddr
+            addrType = B.runGet getHeader (BL.fromStrict addrBytes)
+        disjoin
+          [ addrType === BaseAddress CredentialScriptHash CredentialKeyHash
+          , addrType === BaseAddress CredentialScriptHash CredentialScriptHash
+          , addrType === PointerAddress CredentialScriptHash
+          , addrType === EnterpriseAddress CredentialScriptHash
+          ]
+
+-- To be extra sure, we also test our code with some golden addresses we
+-- generated with "cardano-addresses":
+
+unit_addressType_byronGolden :: Expectation
+unit_addressType_byronGolden =
+    B.runGet getHeader byronAddrGolden `shouldBe` BootstrapAddress
+
+unit_addressType_shelleyEnterprisePaymentGolden :: Expectation
+unit_addressType_shelleyEnterprisePaymentGolden =
+    B.runGet getHeader shelleyEnterprisePaymentAddrGolden
+        `shouldBe` EnterpriseAddress CredentialKeyHash
+
+unit_addressType_stakeAddrGolden :: Expectation
+unit_addressType_stakeAddrGolden =
+    B.runGet getHeader stakeAddrGolden
+        `shouldBe` StakeAddress CredentialKeyHash
+
+unit_addressType_pointerAddrGolden :: Expectation
+unit_addressType_pointerAddrGolden =
+    B.runGet getHeader pointerAddrGolden
+        `shouldBe` PointerAddress CredentialKeyHash
+
+unit_addressType_delegationAddrGolden :: Expectation
+unit_addressType_delegationAddrGolden =
+    B.runGet getHeader delegationAddrGolden
+        `shouldBe` BaseAddress CredentialKeyHash CredentialKeyHash
+
+-- TODO generate more unit tests for each type of address.
```

Comment:
These were the only types of addresses I could generate with cardano-addresses. I would have liked to generate one of each of the 11 different types of addresses. Does anyone know how I might do that?